[Concurrency] Guard use of async calling convention.

varungandhi-apple opened the pull request:
Certain targets don't support the async calling convention, so we first
add the feature check to avoid breaking the codegen/runtime while doing
gradual rollout for different targets.

LLVM patch introducing feature: https://reviews.llvm.org/D96802

aschwaighofer reviewed (changes requested):
This will also need a change to IRGen because we are already unconditionally using `swiftasync` on the async context parameter.

varungandhi-apple commented:
@swift-ci smoke test

aschwaighofer commented on include/swift/Runtime/Config.h:
```diff
@@ -176,15 +176,17 @@ extern uintptr_t __COMPATIBILITY_LIBRARIES_CANNOT_CHECK_THE_IS_SWIFT_BIT_DIRECTL
 #endif
 
-#if __has_attribute(swift_async_context)
+#if __has_feature(swiftasynccc) && USE_SWIFT_ASYNC_LOWERING && \
```
Please remove this part. We can always use `swift_async_context`. We do so today.

varungandhi-apple replied:
Removed.

aschwaighofer commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -2400,7 +2400,8 @@ llvm::Function *IRGenFunction::getOrCreateResumePrjFn() {
         // TODO: remove this once all platforms support lowering the intrinsic.
         // At the time of this writing only arm64 supports it.
-        if (IGM.TargetInfo.canUseSwiftAsyncContextAddrIntrinsic()) {
+        if (IGM.getOptions().UseAsyncLowering &&
```
Same here. We already support/emit `swift_async_context` in tree. This does not need to be condititionalized on `UseAsyncLowering`.

varungandhi-apple replied:
Removed.

varungandhi-apple committed:
[Concurrency] Guard use of async calling convention.

Certain targets don't support the async calling convention, so we first
add the feature check to avoid breaking the codegen/runtime while doing
gradual rollout for different targets.

varungandhi-apple commented:
@swift-ci smoke test

aschwaighofer reviewed (approved):
lgtm

aschwaighofer commented:
@swift-ci please smoke test linux

varungandhi-apple commented:
@swift-ci smoke test and merge
